python/rmm/_lib/device_buffer.pyx

```diff
@@ -133,6 +133,35 @@ cdef class DeviceBuffer:
         }
         return intf
 
+    def copy(self):
+        """Returns a deep copy of DeviceBuffer.
+
+        Returns
+        -------
+        A deep copy of existing ``DeviceBuffer``
+
+        Examples
+        --------
+        >>> import rmm
+        >>> db = rmm.DeviceBuffer.to_device(b"abc")
+        >>> db_copy = db.copy()
+        >>> db.copy_to_host()
+        array([97, 98, 99], dtype=uint8)
+        >>> db_copy.copy_to_host()
+        array([97, 98, 99], dtype=uint8)
+        >>> assert db is not db_copy
+        >>> assert db.ptr != db_copy.ptr
+        """
+        ret = DeviceBuffer.c_from_unique_ptr(
+            make_unique[device_buffer](self.c_obj.get()[0], self.stream.view())
+        )
+        ret.mr = self.mr
+        ret.stream = self.stream
+        return ret
```

Comment:
This is almost possible through the `DeviceBuffer` constructor alone.

PR ( https://github.com/rapidsai/rmm/pull/953 ) would allow the memory resource to be specified as well. Wonder if it makes sense to clean that up and get it in, then update this to use the constructor.

```suggestion
        return DeviceBuffer(ptr=self.ptr, size=self.size, stream=self.stream)
```

Reply:
Looks like @jakirkham's code suggestion returns a deep copy too @shwina WDYT?

Reply:
Oh I see. We want to keep multiple references to the underlying C++ object in some cases? Like using a `shared_ptr`?

Reply:
Am not sure the term deepcopy makes a lot of sense in this context.

Usually the meaning of deepcopy vs. (shallow) copy is clearer when thinking of collections (`list`, `dict`, `set`, etc.) or classes/objects (potentially nested). IOW cases where there is a difference between copying the top-level (new collection or object with the same references) vs. recursing through the structure and copying each object found.

`DeviceBuffer` seems more like `bytes` where there is not really nesting. So deepcopy and (shallow) copy here are the same.

Though maybe I'm missing something (like we care about `shared_ptr` vs. `unique_ptr`).

Reply:
> `DeviceBuffer` seems more like `bytes` where there is not really nesting. So deepcopy and (shallow) copy here are the same.

Got it. Apologies for the confusion, we are then only trying to implement a copy which I suppose can be achieved from the constructor like you pointed.

Reply:
Thanks @jakirkham ! Resolving this as I just addressed the comment.